test/routing/Api.test.ts

```diff
@@ -39,17 +37,27 @@ describe('info api', () => {
                     import_date: expected.import_date,
                     version: expected.version,
                     profiles: [],
+                    elevation: expected.elevation,
                 })
             )
         })
 
+        // this is a little indirect. we are making assertions in the callback of the async api call
+        // if the assertion is wrong the error is caught by the error handler of the api, dispatching an errorAction
+        // therefore we must make the test fail manually
```

Comment:
Couldn't we instead return the promise to the test and make the assert there?

Reply:
There is an async and a synchronous method in the api. We could certainly test the async ones, which would make testing easier. On the other hand our application uses the "withDispatch" methods. We would not test whether the expected actions are dispatched after an api call happens.

Reply:
If we want to test the 'dispatch' calls I think the most natural thing would be to spy on `Dispatcher` rather than actually registering listeners, something like this:

```javascript
// our spy will register all calls to Dispatcher.dispatch
const spy = jest.spyOn(Dispatcher, 'dispatch')
// run our code. we cannot await this because infoWithDispatch is not async
new ApiImpl().infoWithDispatch()
// because infoWithDispatch does an async call but does not return a Promise
// we need to make sure we wait until it is finished... 
// I took this from here: https://github.com/facebook/jest/issues/2157#issuecomment-279171856 we could make this a helper/util function
const flushPromises = () => new Promise(setImmediate)
await flushPromises()
// these are all the calls that were made to Dispatcher.dispatch
console.log(spy.mock.calls)
// we can setup some expectations
expect(spy).toHaveBeenCalled()
expect(spy).toHaveBeenCalledTimes(1)
expect((spy.mock.calls[0][0] as any).result).toEqual(expected)
```

Alternatively we could maybe we return a Promise from `infoWithDispatch` (or use an optional callback?), then we would not need `flushPromises` to make sure our `expect` calls run after the `Dispatcher` was called.

Reply:
Ah, of course. Spying on the dispatcher is much better! Thanks for the suggestion I'll try to go with this approach. 

> Alternatively we could maybe we return a Promise from infoWithDispatch

The ```infoWithDispatch``` methods are there so that they can be called without ```await``` but with no warnings that promises are not handled. Maybe for the sake of better error handling we could disable this warning and make those methods async as well.

Reply:
> The infoWithDispatch methods are there so that they can be called without await but with no warnings that promises are not handled. Maybe for the sake of better error handling we could disable this warning and make those methods async as well.

Or use an optional callback? But unless there are cases where it does not work I think the `flushPromises` thing isn't so bad either.

Reply:
Okay, went for the flush promises approach. What do you think now? If you're happy I will merge this.

Reply:
Thanks, looks good to me. The only thing I noticed is that still no test fails when I put something like `expect(2).toEqual(3)` in `compareRequestBodyAndResolve`.

Reply:
So, when I tested this it worked. The tests fail because the action received by the Dispatcher is not the same as the expected one.

Reply:
Ah yes you are right. I did not realize there are multiple `describe` blocks in `Api.test.ts`.

Reply:
I will merge this then.